daemon/graphdriver/devmapper/deviceset.go

```diff
@@ -2128,13 +2173,13 @@ func (devices *DeviceSet) removeDevice(devname string) error {
 	return err
 }
 
-func (devices *DeviceSet) cancelDeferredRemoval(info *devInfo) error {
+func (devices *DeviceSet) cancelDeferredRemovalIfNeeded(info *devInfo) error {
 	if !devices.deferredRemove {
 		return nil
 	}
 
-	logrus.Debugf("devmapper: cancelDeferredRemoval START(%s)", info.Name())
-	defer logrus.Debugf("devmapper: cancelDeferredRemoval END(%s)", info.Name())
+	logrus.Debugf("devmapper: cancelDeferredRemovalIfNeeded START(%s)", info.Name())
+	defer logrus.Debugf("devmapper: cancelDeferredRemovalIfNeeded END(%s)", info.Name())
```

Comment:
Agreed. We should be able to reuse cancelDeferredRemoval() in cancelDeferredRemovalIfNeeded().

Although not exactly here, we should be calling after.

        if devinfo != nil && devinfo.DeferredRemove == 0 {
                return nil
        }

Also if cancelDeferredRemoval() returns, ErrEnxio, then that means device is gone and this function should still return success as deferred removal is not needed any more.